state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1201,199 +1221,266 @@ pub mod pallet {
-						if swap.output_asset() == Asset::Flip {
-							FlipToBurn::<T>::mutate(|total| {
-								total.saturating_accrue(output_amount);
-							});
-						} else {
-							log_or_panic!(
-								"NetworkFee burning should not be in asset: {:?}",
-								swap.output_asset()
-							);
-						}
-					},
-					SwapRequestState::IngressEgressFee => {
-						Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id });
+					};
 
-						if swap.output_asset() ==
-							ForeignChain::from(swap.output_asset()).gas_asset()
-						{
-							T::IngressEgressFeeHandler::accrue_withheld_fee(
-								swap.output_asset(),
-								output_amount,
+					if is_gas_swap {
+						*gas_swap_state = GasSwapState::OutputReady { gas_budget: output_amount };
+					} else {
+						// The executed swap must be for the principal amount:
+						dca_state.accumulated_output_amount += output_amount;
+
+						// Schedule any remaining DCA chunks:
+						dca_state.scheduled_chunk_swap_id = if dca_state.remaining_chunks > 0 {
+							let chunk_input_amount = dca_state.prepare_next_chunk();
+
+							let swap_id = Self::schedule_swap(
+								request.input_asset,
+								request.output_asset,
+								chunk_input_amount,
+								swap.swap.refund_params.clone(),
+								SwapType::CcmPrincipal,
+								swap_request_id,
+								dca_state.chunk_interval.into(),
 							);
+
+							Some(swap_id)
 						} else {
-							log_or_panic!(
-								"IngressEgressFee swap should not be to non-gas asset: {:?}",
-								swap.output_asset()
+							None
+						};
+
+						// See if we still need to schedule the gas swap
+						if let GasSwapState::ToBeScheduled { gas_budget, other_gas_asset } =
+							gas_swap_state
+						{
+							*gas_swap_state = Self::schedule_ccm_gas_swap(
+								swap_request_id,
+								request.input_asset,
+								*other_gas_asset,
+								*gas_budget,
 							);
 						}
-					},
-				}
+					}
+
+					// Process outcome if there aren't any more scheduled swaps,
+					// otherwise put the request state back:
+					if let (GasSwapState::OutputReady { gas_budget }, None) =
+						(gas_swap_state, dca_state.scheduled_chunk_swap_id)
+					{
+						Self::egress_for_swap(
+							swap_request_id,
+							dca_state.accumulated_output_amount,
+							request.output_asset,
+							output_address.clone(),
+							Some((ccm_deposit_metadata.clone(), *gas_budget)),
+							false, /* refund */
+						);
+
+						true
+					} else {
+						false
+					}
+				},
+				SwapRequestState::Regular { output_address, dca_state } => {
+					if dca_state.remaining_chunks > 0 {
+						let chunk_input_amount = dca_state.prepare_next_chunk();
+						dca_state.accumulated_output_amount += output_amount;
+
+						Self::schedule_swap(
+							request.input_asset,
+							request.output_asset,
+							chunk_input_amount,
+							// TODO: recompute swap params depending on the output amount we got
+							swap.swap.refund_params.clone(),
```

Comment:
I think we already have the distinction between "swap requests" and "swaps" (I chose to continue using "swap" in the context of amm to isloate the changes to the swapping pallet).

In this specific line, the two swaps are both actually at the "execution" level. I agree the naming isn't great, but the outer swap refers to `SwapState` (contains the result of the execution), while the inner is `Swap` the immutable parameters of the execution/swap. I will create a getter method for `refund_params` like I did for other fields to avoid this chaining.